nextcloud-client: miscellaneous fixes

eqyiel opened the pull request:
###### Motivation for this change

I've been maintaining a version of this [outside of nixpkgs](https://github.com/eqyiel/nixos-config/blob/0096911831147821e709e0c07e46ebe09810c12b/pkgs/nextcloud-client/default.nix) for a while and since this is a thing now I would like to contribute the difference to this expression.

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

- enabled parallel building
- use existing cmake* attrs rather than manipulating cmake flags in preConfigure
- use inotify for more efficient monitoring of file changes
- provide gnome-keyring integration when built with `withGnomeKeying = true;`

I'm not sure if adding `gnome_keyring` to `LD_LIBRARY_PATH` is the correct way to go about this, but it's what worked for me.

@caugner thanks for your work on this package.

---

Commit:
nextcloud-client: use cmakeDir and cmakeFlags attrs

Commit:
nextcloud-client: add inotify dependency when building on linux

This allows for more efficient change detection.

Commit:
nextcloud-client: enable parallel building

So that the client can build faster.

bjornfor reviewed (changes requested):
"nextcloud-client: add optional inotify dependency"  -- it doesn't look optional to me. It looks like it will break if "inotify-tools = null". I'm fine with inotify being hard dependency on Linux, but wondering if this expression currently works on Darwin and that adding inotify will break that?

eqyiel commented:
My bad, I'll make it so that it checks `stdenv.isLinux` - what I intended to express is that nextcloud-client tries to find inotify at build time and uses it if it's available.

Happily I'm at my workplace right now where I have access to a Mac to build on, and the build fails with

```
make[2]: Entering directory '/private/var/folders/jt/qgrx8lp10ylbdlnb93t54r6c0000gn/T/nix-build-alsa-lib-1.1.4.1.drv-0/alsa-lib-1.1.4.1/src/control'
  CC       cards.lo
In file included from cards.c:35:
In file included from ./control_local.h:22:
../../include/local.h:47:2: error: Header defining endianness not defined
#error Header defining endianness not defined
 ^
1 error generated.
make[2]: *** [Makefile:334: cards.lo] Error 1
make[2]: Leaving directory '/private/var/folders/jt/qgrx8lp10ylbdlnb93t54r6c0000gn/T/nix-build-alsa-lib-1.1.4.1.drv-0/alsa-lib-1.1.4.1/src/control'
make[1]: *** [Makefile:478: all-recursive] Error 1
make[1]: Leaving directory '/private/var/folders/jt/qgrx8lp10ylbdlnb93t54r6c0000gn/T/nix-build-alsa-lib-1.1.4.1.drv-0/alsa-lib-1.1.4.1/src'
make: *** [Makefile:338: all-recursive] Error 1
builder for ‘/nix/store/9djjg7idiqp7awyx6v3568fkykacrs3l-alsa-lib-1.1.4.1.drv’ failed with exit code 2
fetching path ‘/nix/store/3ip043cbchgj6m4hjfiwn0qw87jdcaz0-qtbase-5.6.2-dev’...
cannot build derivation ‘/nix/store/2chdyipixw3ncqisy3316w98vrwaj9m5-qtmultimedia-5.6.2.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/z6yl0n1s54ny3zfqz660kfx14w8sn3l0-qtlocation-5.6.2.drv’: 1 dependencies couldn't be built
cannot build derivation ‘/nix/store/24axc64rwn91h8yk8hv8b6vlg2r7jgqz-qtwebkit-5.6.2.drv’: 1 dependencies couldn't be built
killing process 7940
cannot build derivation ‘/nix/store/hn2r0a5186ygw93bxzwh7w74y5v0ik9i-nextcloud-client-2.3.2.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/hn2r0a5186ygw93bxzwh7w74y5v0ik9i-nextcloud-client-2.3.2.drv’ failed
nix-build -A nextcloud-client -I nixpkgs=.  5.87s user 4.06s system 54% cpu 18.223 total
```

Which is a dependency of `qtmultimedia` and was not introduced by this commit.

[I'm certain this can build on Darwin](https://github.com/owncloud/client/blob/master/doc/building.rst#mac-os-x), but I think that's out of scope for this PR.

Commit:
nextcloud-client: provide optional gnome-keyring integration

`qtkeyring` can use `gnome-keyring`, but it needs some help to find it.

I have not enabled this by default because not everyone who uses this will want
to pull in GNOME dependencies.

eqyiel commented:
Rebased!

bjornfor commented:
Tested locally. Pushed to master (e828dcb5cd4ae321dde053d2eb3a24992b42f167 and parents) and release-17.09 (e61f1a24c4f654de69286626bcb236d8ca117734 and parents). Thanks!

caugner commented:
Thank you @eqy and @bjornfor!

In general, it would make sense to merge [owncloud-client](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/owncloud-client/default.nix) and [nextcloud-client](https://github.com/NixOS/nixpkgs/blob/master/pkgs/applications/networking/nextcloud-client/default.nix) in the sense that the *nextcloud-client* config could inherit from the *owncloud-client*. Unfortunately, I don't have the Nix experience to accomplish this.
